msl-2.0/docs/man/man3/Render.3
.TH "Render" 3 "24 Jul 2003" "Motion Strategy Library" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Render \- A rendering class that accepts commands from a \fBGui\fP, and determines using specific graphics libraries how to draw the results on a screen. 
.SH SYNOPSIS
.br
.PP
\fC#include <render.h>\fP
.PP
Inherited by \fBRenderGL\fP, \fBRenderIv\fP, and \fBRenderPerformer\fP.
.PP
.SS "Public Methods"

.in +1c
.ti -1c
.RI "\fBRender\fP ()"
.br
.ti -1c
.RI "\fBRender\fP (string filepath)"
.br
.ti -1c
.RI "\fBRender\fP (\fBScene\fP *s, string filepath)"
.br
.ti -1c
.RI "virtual \fB~Render\fP ()"
.br
.ti -1c
.RI "virtual void \fBInit\fP ()"
.br
.RI "\fIInitialized the renderer.\fP"
.ti -1c
.RI "virtual void \fBHandleEvents\fP ()"
.br
.RI "\fIProcess IO events.\fP"
.ti -1c
.RI "virtual void \fBMainLoop\fP (\fBGui\fP *g)"
.br
.RI "\fIIf ControlFreak = true, then MainLoop is entered here.\fP"
.ti -1c
.RI "virtual void \fBReset\fP ()"
.br
.RI "\fIReset the renderer.\fP"
.ti -1c
.RI "virtual void \fBTerminate\fP ()"
.br
.RI "\fIImplement functions upon termination of the renderer.\fP"
.ti -1c
.RI "virtual void \fBSetFrameList\fP ()"
.br
.RI "\fIMake the set of frames from StateList and TimeList.\fP"
.ti -1c
.RI "virtual void \fBMakeAnimationFrames\fP (const list< \fBMSLVector\fP > &xlist, double deltat)"
.br
.RI "\fIGenerate FrameList and set AnimationActive to true.\fP"
.ti -1c
.RI "virtual void \fBMakeAnimationFrames\fP (const list< \fBMSLVector\fP > &xlist, const list< double > &timelist)"
.br
.ti -1c
.RI "virtual void \fBDrawPath\fP ()"
.br
.RI "\fIDisplay an entire path (the specific renderer determines how).\fP"
.ti -1c
.RI "virtual void \fBButtonHandle\fP (int b)"
.br
.RI "\fIExecute actions for render control window choices.\fP"
.ti -1c
.RI "void \fBSetScene\fP (\fBScene\fP *s)"
.br
.RI "\fIChange the associated scene.\fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "string \fBFilePath\fP"
.br
.RI "\fIThe path name for accessing files.\fP"
.ti -1c
.RI "list< \fBMSLVector\fP > \fBFrameList\fP"
.br
.RI "\fIThe animation frames; each element is a SceneConfiguration.\fP"
.ti -1c
.RI "double \fBFrameTime\fP"
.br
.RI "\fIThe amount of time for which a frame is shown.\fP"
.ti -1c
.RI "float \fBLastFrameTime\fP"
.br
.RI "\fIThe time stamp of the last frame change.\fP"
.ti -1c
.RI "float \fBFrameStuckTime\fP"
.br
.RI "\fIThe amount of time since the last frame change.\fP"
.ti -1c
.RI "int \fBNumFrames\fP"
.br
.RI "\fIThe number of frames in the animation.\fP"
.ti -1c
.RI "double \fBAnimationStartPause\fP"
.br
.RI "\fINumber of seconds to wait at the start of an animation.\fP"
.ti -1c
.RI "double \fBAnimationEndPause\fP"
.br
.RI "\fINumber of seconds to wait at the end of an animation.\fP"
.ti -1c
.RI "double \fBAnimationTimeScale\fP"
.br
.RI "\fIThe speedup factor for the animation (1.0 = normal speed).\fP"
.ti -1c
.RI "double \fBrealAnimationTimeScale\fP"
.br
.ti -1c
.RI "double \fBAnimationTimeScale_prev\fP"
.br
.ti -1c
.RI "double \fBrealAnimationTimeScale_prev\fP"
.br
.ti -1c
.RI "int \fBAnimationFrameIndex\fP"
.br
.RI "\fIThe index in FrameList of the frame that should be currently shown.\fP"
.ti -1c
.RI "\fBMSLVector\fP \fBCurrentAnimationFrame\fP"
.br
.RI "\fIThe frame that should be shown currently.\fP"
.ti -1c
.RI "bool \fBAnimationActive\fP"
.br
.RI "\fISet to true to start the animation.\fP"
.ti -1c
.RI "bool \fBControlFreak\fP"
.br
.RI "\fISet to true if the renderer needs main loop control; otherwise, iterated polling of HandleEvents will be performed automatically.\fP"
.ti -1c
.RI "bool \fBRenderCtlWindowOn\fP"
.br
.RI "\fIIs teh rendering control window currently displayed?\fP"
.ti -1c
.RI "bool \fBAttachedCameraOn\fP"
.br
.RI "\fISet to true for the viewpoint to be attached to a body.\fP"
.ti -1c
.RI "bool \fBBoundingBoxOn\fP"
.br
.RI "\fISet to true to draw a bounding box using S->LowerWorld and S->UpperWorld.\fP"
.ti -1c
.RI "bool \fBMultipleViewsOn\fP"
.br
.RI "\fISet to true to show multiple views simultaneously.\fP"
.ti -1c
.RI "bool \fBShowPathOn\fP"
.br
.RI "\fISet to true to show the entire path.\fP"
.ti -1c
.RI "double \fBAmbientLight\fP"
.br
.RI "\fIA parameter that controls the amount of light.\fP"
.in -1c
.SS "Protected Methods"

.in +1c
.ti -1c
.RI "virtual void \fBSetCurrentAnimationFrame\fP ()"
.br
.RI "\fIThis uses the FrameList to return the SceneConfiguration that is supposed to be shown at the present time.\fP"
.ti -1c
.RI "virtual void \fBShowCurrentAnimationFrame\fP ()"
.br
.RI "\fIDisplay the current animation frame in a rendering window.\fP"
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "\fBScene\fP * \fBS\fP"
.br
.RI "\fIAllow information from a \fBScene\fP to be accessed.\fP"
.ti -1c
.RI "list< \fBMSLVector\fP > \fBStateList\fP"
.br
.RI "\fIA sequence of state to use for animation (should be generated by \fBGui\fP).\fP"
.ti -1c
.RI "list< double > \fBTimeList\fP"
.br
.RI "\fIThe time stamps for the state sequence.\fP"
.ti -1c
.RI "list< string > \fBEnvList\fP"
.br
.ti -1c
.RI "list< string > \fBBodyList\fP"
.br
.ti -1c
.RI "float \fBRGBRed\fP [RENDERCOLORS]"
.br
.RI "\fIRGB color values.\fP"
.ti -1c
.RI "float \fBRGBGreen\fP [RENDERCOLORS]"
.br
.ti -1c
.RI "float \fBRGBBlue\fP [RENDERCOLORS]"
.br
.in -1c
.SH "DETAILED DESCRIPTION"
.PP 
A rendering class that accepts commands from a \fBGui\fP, and determines using specific graphics libraries how to draw the results on a screen.
.PP
This hierarchy of classes contains different implementations of graphical rendering requests. For example, when a graphical user interface (GUI) requests that the a solution path is animated, a method in a Render class displays the bodies in motion using configurations obtained from the \fBScene\fP class. Each derived class in Render corresponds to a different graphics system. Presently, there are renderers for Open Inventor and Open GL. The flexibility provided by these classes enables easy extensions to be made for other graphics libraries and platforms, such as Open Inventor.
.PP
The rendering is expressed in terms of a \fBScene\fP, in which the  scene configuration gives the configuration of a collection of Bodies in a static Environment. 
.PP
.SH "CONSTRUCTOR & DESTRUCTOR DOCUMENTATION"
.PP 
.SS "Render::Render ()"
.PP
.SS "Render::Render (string filepath)"
.PP
.SS "Render::Render (\fBScene\fP * s, string filepath)"
.PP
.SS "virtual Render::~Render ()\fC [inline, virtual]\fP"
.PP
.SH "MEMBER FUNCTION DOCUMENTATION"
.PP 
.SS "void Render::ButtonHandle (int b)\fC [virtual]\fP"
.PP
Execute actions for render control window choices.
.PP
.SS "virtual void Render::DrawPath ()\fC [inline, virtual]\fP"
.PP
Display an entire path (the specific renderer determines how).
.PP
Reimplemented in \fBRenderGL\fP.
.SS "virtual void Render::HandleEvents ()\fC [inline, virtual]\fP"
.PP
Process IO events.
.PP
Reimplemented in \fBRenderPerformer\fP.
.SS "void Render::Init ()\fC [virtual]\fP"
.PP
Initialized the renderer.
.PP
Reimplemented in \fBRenderGL\fP.
.SS "void Render::MainLoop (\fBGui\fP * g)\fC [virtual]\fP"
.PP
If ControlFreak = true, then MainLoop is entered here.
.PP
Reimplemented in \fBRenderGL\fP.
.SS "void Render::MakeAnimationFrames (const list< \fBMSLVector\fP > & xlist, const list< double > & timelist)\fC [virtual]\fP"
.PP
.SS "void Render::MakeAnimationFrames (const list< \fBMSLVector\fP > & xlist, double deltat)\fC [virtual]\fP"
.PP
Generate FrameList and set AnimationActive to true.
.PP
.SS "void Render::Reset ()\fC [virtual]\fP"
.PP
Reset the renderer.
.PP
Reimplemented in \fBRenderGL\fP.
.SS "void Render::SetCurrentAnimationFrame ()\fC [protected, virtual]\fP"
.PP
This uses the FrameList to return the SceneConfiguration that is supposed to be shown at the present time.
.PP
.SS "void Render::SetFrameList ()\fC [virtual]\fP"
.PP
Make the set of frames from StateList and TimeList.
.PP
.SS "void Render::SetScene (\fBScene\fP * s)"
.PP
Change the associated scene.
.PP
.SS "virtual void Render::ShowCurrentAnimationFrame ()\fC [inline, protected, virtual]\fP"
.PP
Display the current animation frame in a rendering window.
.PP
Reimplemented in \fBRenderPerformer\fP.
.SS "virtual void Render::Terminate ()\fC [inline, virtual]\fP"
.PP
Implement functions upon termination of the renderer.
.PP
Reimplemented in \fBRenderPerformer\fP.
.SH "MEMBER DATA DOCUMENTATION"
.PP 
.SS "double Render::AmbientLight"
.PP
A parameter that controls the amount of light.
.PP
.SS "bool Render::AnimationActive"
.PP
Set to true to start the animation.
.PP
.SS "double Render::AnimationEndPause"
.PP
Number of seconds to wait at the end of an animation.
.PP
.SS "int Render::AnimationFrameIndex"
.PP
The index in FrameList of the frame that should be currently shown.
.PP
.SS "double Render::AnimationStartPause"
.PP
Number of seconds to wait at the start of an animation.
.PP
.SS "double Render::AnimationTimeScale"
.PP
The speedup factor for the animation (1.0 = normal speed).
.PP
.SS "double Render::AnimationTimeScale_prev"
.PP
.SS "bool Render::AttachedCameraOn"
.PP
Set to true for the viewpoint to be attached to a body.
.PP
.SS "list<string> Render::BodyList\fC [protected]\fP"
.PP
.SS "bool Render::BoundingBoxOn"
.PP
Set to true to draw a bounding box using S->LowerWorld and S->UpperWorld.
.PP
.SS "bool Render::ControlFreak"
.PP
Set to true if the renderer needs main loop control; otherwise, iterated polling of HandleEvents will be performed automatically.
.PP
.SS "\fBMSLVector\fP Render::CurrentAnimationFrame"
.PP
The frame that should be shown currently.
.PP
.SS "list<string> Render::EnvList\fC [protected]\fP"
.PP
.SS "string Render::FilePath"
.PP
The path name for accessing files.
.PP
.SS "list<\fBMSLVector\fP> Render::FrameList"
.PP
The animation frames; each element is a SceneConfiguration.
.PP
.SS "float Render::FrameStuckTime"
.PP
The amount of time since the last frame change.
.PP
.SS "double Render::FrameTime"
.PP
The amount of time for which a frame is shown.
.PP
.SS "float Render::LastFrameTime"
.PP
The time stamp of the last frame change.
.PP
.SS "bool Render::MultipleViewsOn"
.PP
Set to true to show multiple views simultaneously.
.PP
.SS "int Render::NumFrames"
.PP
The number of frames in the animation.
.PP
.SS "double Render::realAnimationTimeScale"
.PP
.SS "double Render::realAnimationTimeScale_prev"
.PP
.SS "bool Render::RenderCtlWindowOn"
.PP
Is teh rendering control window currently displayed?
.PP
.SS "float Render::RGBBlue[RENDERCOLORS]\fC [protected]\fP"
.PP
.SS "float Render::RGBGreen[RENDERCOLORS]\fC [protected]\fP"
.PP
.SS "float Render::RGBRed[RENDERCOLORS]\fC [protected]\fP"
.PP
RGB color values.
.PP
.SS "\fBScene\fP* Render::S\fC [protected]\fP"
.PP
Allow information from a \fBScene\fP to be accessed.
.PP
.SS "bool Render::ShowPathOn"
.PP
Set to true to show the entire path.
.PP
.SS "list<\fBMSLVector\fP> Render::StateList\fC [protected]\fP"
.PP
A sequence of state to use for animation (should be generated by \fBGui\fP).
.PP
.SS "list<double> Render::TimeList\fC [protected]\fP"
.PP
The time stamps for the state sequence.
.PP


.SH "AUTHOR"
.PP 
Generated automatically by Doxygen for Motion Strategy Library from the source code.


msl-2.0/src_win/render.cpp
//----------------------------------------------------------------------
//               The Motion Strategy Library (MSL)
//----------------------------------------------------------------------
//
// Copyright (c) University of Illinois and Steven M. LaValle.     
// All Rights Reserved.
// 
// Permission to use, copy, and distribute this software and its 
// documentation is hereby granted free of charge, provided that 
// (1) it is not a component of a commercial product, and 
// (2) this notice appears in all copies of the software and
//     related documentation. 
// 
// The University of Illinois and the author make no representations
// about the suitability or fitness of this software for any purpose.  
// It is provided "as is" without express or implied warranty.
//----------------------------------------------------------------------

#include <math.h>
#include <stdlib.h>

#include "render.h"
#include "defs.h"

// Include the VCR buttons
#include "pixmaps/ff.xpm"
#include "pixmaps/pause.xpm"
#include "pixmaps/play.xpm"
#include "pixmaps/rw.xpm"
#include "pixmaps/stop.xpm"
#include "pixmaps/prev.xpm"
#include "pixmaps/next.xpm"

// Include the msl icon
#include "pixmaps/msl.xpm"


////////////////////////////////////////////////////////
//
// Render Class
//
///////////////////////////////////////////////////////

Render::Render() {
  Render("");
}


Render::Render(string filepath)
{
  FilePath = filepath;
}


Render::Render(Scene *s, string filepath)
{
  SetScene(s);

  FilePath = filepath;
  ControlFreak = false;
}


// This sets up the file names for the environment and movable bodies
void Render::Init()
{
  int i;
  std::ifstream fin;

  Reset();

  FrameTime = 0.1; // This is default; each renderer should determine this
  AnimationStartPause = 0.0; // should be in seconds
  AnimationEndPause = 0.0;
  RenderCtlWindowOn = false;

  AttachedCameraOn = false;
  BoundingBoxOn = false;
  MultipleViewsOn = false;
  ShowPathOn = false;

  // Colors for objects (add more if you like, but increase RENDERCOLORS)
  //RGBRed[0] = .467; RGBGreen[0] = 0.53; RGBBlue[0] = 0.6; // Light Slate Gray
  RGBRed[0] = 1.0; RGBGreen[0] = 0.0; RGBBlue[0] = 1.0; // Magenta
  RGBRed[1] = 0.8; RGBGreen[1] = 0.0; RGBBlue[1] = 0.0; // Red
  RGBRed[2] = 0.0; RGBGreen[2] = 0.7; RGBBlue[2] = 0.0; // Green
  RGBRed[3] = 0.7; RGBGreen[3] = 0.7; RGBBlue[3] = 0.0; // Yellow
  RGBRed[4] = 0.0; RGBGreen[4] = 0.7; RGBBlue[4] = 0.7; // Lightblue
  RGBRed[5] = 1.0; RGBGreen[5] = 0.0; RGBBlue[5] = 1.0; // Magenta
  RGBRed[6] = 0.0; RGBGreen[6] = 0.0; RGBBlue[6] = 0.7; // Blue
  RGBRed[7] = 1.0; RGBGreen[7] = 0.65; RGBBlue[7] = 0.0; // Orange
  RGBRed[8] = 1.0; RGBGreen[8] = 0.078; RGBBlue[8] = 0.576; // DeepPink
  RGBRed[9] = 1.0; RGBGreen[9] = 0.51; RGBBlue[9] = 0.278; // Sienna1

  fin.clear();
  fin.open((FilePath + "EnvList").c_str());
  if (fin)
    fin >> EnvList;
  else {
    fin.clear();
    fin.open((FilePath + "Obst").c_str());
    if (fin)
      EnvList.push_back("Obst");
  }
  fin.close();

  //cout << "EnvList: " << EnvList.size() << endl;

  char* s = new char[50];

  // Load bodies
  fin.clear();
  fin.open((FilePath+"BodyList").c_str());
  if (fin)
    fin >> BodyList;
  else {
    fin.clear();
    fin.open((FilePath+"Robot").c_str());
    if (fin)
      BodyList.push_back("Robot");
    else { // Multiple robots 
      for (i = 0; i < S->NumBodies; i++) {
	    sprintf(s,"%sRobot%d",FilePath.c_str(),i);
	    fin.clear();
		fin.open(s);
	    if (fin) {
	      sprintf(s,"Robot%d",i);
	      BodyList.push_back(string(s)); // Use the robot from World
		}
		fin.close();
      }
    }
  }

  // Close ifstream
  fin.close();

  //cout << "BodyList: " << BodyList.size() << endl;
}



void Render::SetScene(Scene *s) 
{
  S = s; // S is the Scene for Render
}


void Render::MakeAnimationFrames(const list<MSLVector> &xlist, double deltat) 
{
  double t;
  int i;

  StateList = xlist;
  TimeList.clear();
  
  t = 0.0;
  for (i = 0; i < (int) xlist.size(); i++) {
    TimeList.push_back(t);
    t += deltat;
  }
  
  SetFrameList();
}


void Render::MakeAnimationFrames(const list<MSLVector> &xlist,
				 const list<double> &timelist) 
{
  StateList = xlist;
  TimeList = timelist;
  SetFrameList();
}


void Render::SetFrameList() {
  double ltime; // Lower time
  double utime; // Upper time
  double crtime; // Current time
  double lasttime;
  double lambda;
  MSLVector sc; // scene configuration
  MSLVector lx,ux; // lower and upper states
  int i;
  list<MSLVector>::iterator stli,tstli;
  list<double>::iterator tli,li;

  //cout << "Running SetFrameList\n";
  //cout << "StateList: " << StateList << "\n";
  //cout << "TimeList: " << TimeList << "\n";

  FrameList.clear();
  NumFrames = 0;

  if (TimeList.size() < 2) {
    if (TimeList.size() == 1) {
      FrameList.push_back(S->StateToSceneConfiguration(StateList.front()));
      NumFrames = 1;
    }
    return;
  }

  // Make the starting pause
  for (i = 0; i < (int) (AnimationStartPause/FrameTime); i++) {
    FrameList.push_back(S->StateToSceneConfiguration(StateList.front()));
    NumFrames++;
  }

  li = TimeList.begin(); li++;
  stli = StateList.begin(); stli++;
  utime = *li;  // Time index of second element
  ux = *stli;
  ltime = TimeList.front();
  lx = StateList.front();
  crtime = ltime;
  lasttime = TimeList.back();
  tli = li; tli++;
  while (crtime <= lasttime) {
    while ((utime < crtime)&&  // Make sure the crtime within upper and lower
	   (tli != TimeList.end())) {
      li++;
      stli++;
      utime = *li;
      ux = *stli;
      tli = li; tli++;
    }
    tli = li; tli--;
    ltime = *tli;
    tstli = stli; tstli--;
    lx = *tstli;
    while (ltime > crtime) { // This can happen sometimes (I think)!
      tli--;
      tstli--;
      ltime = *tli;
      lx = *tstli;
    }

    lambda = (crtime - ltime)/(utime - ltime);
    //cout << "lambda: " << lambda << "\n";

    // Interoplate based on states lx and ux
    sc = S->InterpolatedSceneConfiguration(lx,ux,lambda);
    FrameList.push_back(sc);
    NumFrames++;
    crtime += FrameTime;
  }
  
  // Add in the last frame, if necessary
  if (FrameList.back() != (S->StateToSceneConfiguration(StateList.back())))
    FrameList.push_back(S->StateToSceneConfiguration(StateList.back()));

  // Make the ending pause
  for (i = 0; i < (int) (AnimationEndPause/FrameTime); i++) {
    FrameList.push_back(S->StateToSceneConfiguration(StateList.back())); 
    NumFrames++;
  }

  //list<MSLVector>::iterator x;
  //cout << "StateList:\n";
  //forall(x,StateList)
  //  cout << *x << "\n";
  //cout << "TimeList: " << TimeList << "\n";
  //cout << "\n\nFrameList:\n";
  //forall(x,FrameList)
  //  cout << *x << "\n";
  //cout << "NumFrames: " << NumFrames << "\n";
}



void Render::SetCurrentAnimationFrame() {
  MSLVector c(S->SceneConfigurationDim);
  int num,skip,i;

  // How long has the frame been stuck?
  FrameStuckTime = used_time() - LastFrameTime;

  skip = (int) (AnimationTimeScale * FrameStuckTime / FrameTime);

  if (skip > 0) {
    AnimationFrameIndex += skip;
    used_time(LastFrameTime);
  }
    
  num = FrameList.size();
  if (AnimationFrameIndex > num - 1) {
    AnimationFrameIndex = 0;
    used_time(LastFrameTime);
  }

  if (AnimationFrameIndex < 0) {
    AnimationFrameIndex = 0;
    used_time(LastFrameTime);
  }

  if (num > 0) {
    list<MSLVector>::iterator fi;
    fi = FrameList.begin();
    for (i = 0; i < AnimationFrameIndex - 1; i++)
      fi++;
    c=*fi;
  }

  CurrentAnimationFrame = c;
}



void Render::ButtonHandle(int b) 
{
  int i;
  list<MSLVector>::iterator fi;
  //cout << "Button " << b << "\n";

  switch (b) {

  case GID_VCR_PLAY: // Play
    if (!AnimationActive) {
      used_time(LastFrameTime);
      AnimationActive = true;
    }
    break;
  case GID_VCR_STOP: // Stop
      AnimationActive = false;
      AnimationFrameIndex = 0;
      LastFrameTime = 0.0;;
      used_time(LastFrameTime);
    break;
  case GID_VCR_PAUSE: // Pause
    if (AnimationActive) {
      AnimationActive = false;
    }
    else {
      AnimationActive = true;
      used_time(LastFrameTime);
    }
    cout << "Frame: " << AnimationFrameIndex << "   Time stamp: " 
	 << AnimationFrameIndex*FrameTime << "s\n";
    break;
  case GID_VCR_SLOWER: // Deccelerate animation
    AnimationTimeScale /= 1.41421356237; // Two presses doubles the speed
    break;
  case GID_VCR_FASTER: // Accelerate animation
    AnimationTimeScale *= 1.41421356237; // Two presses doubles the speed
    break;
  case GID_VCR_RESET:
    Reset();
    break;
  case GID_VCR_LAST: // Prev frame
    AnimationActive = false;
    if (AnimationFrameIndex > 0)
      AnimationFrameIndex--;
    else
      AnimationFrameIndex = FrameList.size() - 1;
    fi = FrameList.begin();
    for (i = 0; i < AnimationFrameIndex - 1; i++)
      fi++;
    CurrentAnimationFrame = *fi;
    ShowCurrentAnimationFrame();
    cout << "Frame: " << AnimationFrameIndex << "   Time stamp: " 
	 << AnimationFrameIndex*FrameTime << "s\n";
   break;
  case GID_VCR_NEXT: // Next frame
    AnimationActive = false;
    if (AnimationFrameIndex < (int) FrameList.size() - 1)
      AnimationFrameIndex++;
    else
      AnimationFrameIndex = 0;
    fi = FrameList.begin();
    for (i = 0; i < AnimationFrameIndex - 1; i++)
      fi++;
    CurrentAnimationFrame = *fi;
    ShowCurrentAnimationFrame();
    cout << "Frame: " << AnimationFrameIndex << "   Time stamp: " 
	 << AnimationFrameIndex*FrameTime << "s\n";
    break;
  case GID_TOGGLE_SHOWPATH: cout << "Toggle Show Path\n";
      ShowPathOn = !ShowPathOn; 
      break;
  case GID_TOGGLE_ATTACHEDCAMERA: cout << "Toggle Attached Camera\n";
      AttachedCameraOn = !AttachedCameraOn; 
      break;
  case GID_TOGGLE_MULTIPLEVIEWS: cout << "Toggle Multiple Views\n";
      MultipleViewsOn = !MultipleViewsOn; 
      break;
  case GID_TOGGLE_BOUNDINGBOX: cout << "Toggle Bounding Box\n";
      BoundingBoxOn = !BoundingBoxOn; 
      break;
  default: cout << "Option " << b << " not implemented\n";
    break;
  }
}


void Render::MainLoop(Gui *g) {

  g->Finished = false;
  while (!g->Finished) {
    g->HandleEvents();
    HandleEvents();
  }

}



void Render::Reset() {
  AnimationActive = false;
  AnimationFrameIndex = 0;
  if (FrameList.size() > 0)
    CurrentAnimationFrame = FrameList.front();
  else
    CurrentAnimationFrame = MSLVector(S->SceneConfigurationDim);
  AnimationTimeScale = 1.0;
  LastFrameTime = 0.0;
  used_time(LastFrameTime);
  // Leave these next four alone because the GUI is not connected to them
  //AttachedCameraOn = false;
  //BoundingBoxOn = false;
  //MultipleViewsOn = false;
  //ShowPathOn = false;
  AmbientLight = 0.2;
}




